asf_heat_pump_suitability/research/exploratory/build_classification_model_block_of_flats.py

```diff
@@ -0,0 +1,798 @@
+    )["avg_n_stacked_uprns"].median()
+
+    # Get the geometry of the example building
+    building_id = typical_examples[f"{label}_building_id"]
+    examples_gdf = buildings_w_uprns_gdf[
+        buildings_w_uprns_gdf["oct25_building_id"] == building_id
+    ][["geometry"]]
+    examples_gdf["colour"] = "blue"
+
+    # Get the UPRNs of the example building
+    temp_examples_gdf = all_uprns_gdf[
+        all_uprns_gdf["oct25_building_id"] == building_id
+    ][["geometry"]]
+    temp_examples_gdf["colour"] = "red"
+
+    # Plot the above info
+    examples_gdf = pd.concat([examples_gdf, temp_examples_gdf])
+    examples_gdf.plot(ax=ax, color=examples_gdf["colour"])
+    ax.set_title(
+        f"{label}, N flats: {typical_examples[f'{label}_n_flats_for_selected_building']}"
+    )
+
+fig.suptitle(
+    "Typical distribution of UPRNs in building footprint per archetype\n(based on median concave hull (m2))"
+)
+fig.tight_layout()
+
+# %%
+# Plot the distribution of features for blocks and non blocks
+nrows = 3
+ncols = math.ceil(len(features) / nrows)
+
+fig, axs = plt.subplots(nrows, ncols, figsize=(15, 8))
+
+for ax, feature in zip(axs.ravel(), features):
+    ax.hist(model_df.filter(pl.col("block_of_flats"))[feature], bins=40, alpha=0.5)
+    ax.hist(model_df.filter(~pl.col("block_of_flats"))[feature], bins=40, alpha=0.5)
+    ax.set_title(feature)
+
+fig.legend(["Block of flats", "Non block"], loc="upper right")
+fig.suptitle("Distribution of features in blocks and non-blocks")
+fig.tight_layout()
+
+# %%
+# Same but with violin plots
+nrows = 3
+ncols = math.ceil(len(features) / nrows)
+
+fig, axs = plt.subplots(nrows, ncols, figsize=(15, 8))
+
+for ax, feature in zip(axs.ravel(), features):
+    vplot = ax.violinplot(
+        dataset=(
+            model_df.filter(pl.col("block_of_flats"))[feature],
+            model_df.filter(~pl.col("block_of_flats"))[feature],
+        )
+    )
+    for i, pc in enumerate(vplot["bodies"], 1):
+        if i % 2 != 0:
+            pc.set_facecolor("gold")
+        else:
+            pc.set_facecolor("royalblue")
+        pc.set_edgecolor("black")
+
+    ax.set_title(feature)
+
+fig.legend(["Block of flats", "Non block"], loc="upper right")
+fig.suptitle("Distribution of features in blocks and non-blocks")
+fig.tight_layout()
+
+# %% [markdown]
+# ## Train and evaluate model and tune hyperparameters
+
+# %%
+# Set random state int and RandomState instance
+random_state = 8
+rng = np.random.RandomState(random_state)
+
+# %%
+# Model features
+features = [
+    "n_UPRNs",
+    "n_flats",
+    "building_area_m2",
+    "building_perimeter_m",
+    "proportion_flats",
+    "UPRNs_per_building_m2",
+    "concave_hull_area_m2",
+    "uprns_per_hull_area_m2",
+    "flats_per_hull_area_m2",
+    "avg_n_stacked_uprns",
+    "std_n_stacked_uprns",
+]
+
+# %%
+# Create param distributions for hyperparameter search
+param_distributions = {
+    "n_estimators": list(range(100, 1000, 10)),
```

Comment:
Unsure if there will be much difference between 100 and 110 trees? So we might want to increase the step here? to 50 or 100.

```suggestion
    "n_estimators": list(range(100, 1000 + 1, 50)),
```